docs/design-proposals/node/Unified-Vertical-Scaling-for-Container-and-VM-Pods.md

```diff
@@ -319,3 +341,129 @@ Other components:
+
+  Two new APIs, SetVcpus() and SetCurrentMemory() were added to support VM vertical scaling, which extends the call to 
+  Libvirt APIs to update the VM domain xml.
+
+	// Update vcpu for a give domain
+	SetVcpus(int) error
+	// Update current memory for a given domain
+	SetCurrentMemory(int64) error
+	
+  Both of them has similar workflow a below:
+
+    1. Get the linux resource from the UpdateContainerResource request
+    2. Get the current configuration from the VM 
+    3. Calculate the delta and determine Attach or Detach operation ( for CPU it will be enable and disable operation)
+    4. Call the Libvirt API to update the domain.
+    
+  The ContainerStatus() implementation is also changed to return the current configured resources in the VM to Kubelet. Only
+  change for the vertical scale feature is to sync the VM domain configuration with the runtime metadata in containerInfo.Config
+  data structure.
+
+#### Type conversions
+  There are a few data struct that handles, transports or stores the resource info as metadata in Kubernetes, VM runtime 
+  and Libvirt. It is important to list here so during conversion the data will not be truncated.
+
+|item	|Kubelet|linuxContainerResource in CRI|LinuxResource & CG	| Runtime vmConfig	|Runtime DomainSettings	|Libvirt domain |
+|-------|-------|-----------------------------|-----------------------|-----------------------|-----------------------|---------------|
+|vcpus	|int64	|-	                      |-	              |-	              |int	              |domainVcpu.Value, int
+|cpuShares|-	|int64, default 0	      |LinuxCPU.Shares, *uint64|Int64	              |uint |domainCpuTuneShares.Value |uint
+|cpuQuota |-	|Int64, default 0.            |LinuxCPU.Quota, *int64	|Int64	|int64	|domainCPUTTuneQuota.Value |int64
+|cpuPeriod|-	|Int64, default 0.            |LinuxCPU.Period, *uint64	|Int64	|Uint64	|domainCPUTunePeriod.Value |uint64
+|Memory	| Int64	|Int64 InBytes, default 0.    |LinuxMemory.Limit *int64, inBytes|Int64	|Int, with Unit def	|domainMemory.Value uint
+
+#### Keep metadata in-sync
+The Arktos-vm-runtime service has to handle three pieces metadata, VM CGroup, vmConfig, and Libvirt domain definition, 
+and ensure they are synchronized when they were used.
+
+All three of them were persisted in different stores, CGroup in files, vmConfig in runtime's metadata store and domain 
+definition in the Libvirt metadata store. it is runtime's responsibility to keep them in sync. When VM resource is 
+created or updated, CGroup, vmConfig, domain definition are persisted in such order:
+
+1. Update CGroup
+2. Update VMResources( vmConfig and domainDef)
+
+If update VMResources operation failed, then the CGroup update operation will be rolled back and error will be raised.
+
+During UpdateVmResources(), no rollback is enforced. The ContainerStatus() function will sync up the vmConfig and 
+Libvirt domain definition if they are out of sync.
```

Comment:
why not do rollback? If these info are out of sync for some period of time, wouldn't it incur some side effect on some operation?

Reply:
You have a point. I am working on the handler design. will add rollback CG update IF the domain update failed. some details of the reasons for the current design.

First of all, ideally the 3 operations ( CG update, libvirt domain update, runtime config metadata update ) should be in one transaction.

The currently approach is "eventually consistency" mainly due to the downstream behavior of update libvirt domain. resource change ( especially decrease mem ) are done asynchronously and the actual status is available in detachDevice event handler.  Note that the runtime metadata will not change if there is any error returned from the domain update. also, the synchronization logic took the domain definition as the source of truth if there is any differences between libvirt domain and vm runtime metadata.